src/main/java/org/eclipse/yasson/internal/serializer/ObjectSerializer.java

```diff
@@ -60,15 +61,24 @@ public ObjectSerializer(CurrentItem<?> wrapper, Type runtimeType, ClassModel cla
 
     @Override
     protected void serializeInternal(T object, JsonGenerator generator, SerializationContext ctx) {
-        final PropertyModel[] allProperties = ((Marshaller) ctx).getMappingContext().getOrCreateClassModel(object.getClass())
-                .getSortedProperties();
-        for (PropertyModel model : allProperties) {
-            try {
-                marshallProperty(object, generator, ctx, model);
-            } catch (Exception e) {
-                throw new JsonbException(Messages.getMessage(MessageKeys.SERIALIZE_PROPERTY_ERROR, model.getWriteName(),
-                                                             object.getClass().getCanonicalName()), e);
+        Marshaller context = (Marshaller) ctx;
+        try {
+            if (context.addProcessedObject(object)) {
```

Comment:
I'm not sure if this will work in all cases. For example, what would happen if an object has two separate references to the same object?
For example:
```java
public static class A {
  public B ref1;
  public B ref2;
}
public static class B {
  public String prop = "foo";
}

A a = new A();
B b = new B();
a.ref1 = b;
b.ref2 = b;
```
If we serialize `a`, it should be:
```json
{ 
  "ref1": { 
    "prop": "foo"
  },
  "ref2": {
    "prop": "foo"
  }
}
```
but with this PR I think we would get:
```json
{ 
  "ref1": { 
    "prop": "foo"
  }
}
```

Reply:
@aguibert thanks for taking a look into this. It is interesting to add this test case, so I have created it in a new test case with your example.

The result is the expected, because before it tries to serialize a.ref2, the a.ref1 was already removed from the processed object hashset.

Reply:
thanks for adding the test @jbescos, I'd still like to add a few more test scenarios before merging this PR. Mainly I want to make sure we don't skip serialization of duplicate objects that are not recursive.

* test a chain of duplicate objects:
```java
Foo foo = new Foo("foo");
Chain c1 = new Chain("c1");
Chain c2 = new Chain("c2");
c1.linksTo = c2;
c1.has = foo;
c2.has = foo;
jsonb.toJson(c1);
```
Should result in:
```json
{ 
  "name": "c1",
  "has" : { "bar": "foo" },
  "linksTo": {
    "name": "c2",
    "has" : { "bar": "foo" }
  }
}
```
I think this scenario will break because we will add `foo` as a processed object when we serialize `c1`, and then when the code traverses down to the child object `c2` it hasn't hit the finally block yet which removes `foo` as a processed object.

* Test the above object setup with an Adapter, because the `AdaptedObjectSerializer` also calls `add/removeProcessedObject()` and I wonder if they will interfere with each other
* Test the above object setup with a Serializer, for the same reason as the Adapter

Reply:
@aguibert I have tested with default jsonb and using the AdaptedObjectSerializer and it works as expected.

But I had to deep inside in the code and now I have a better understanding.

There are 2 serializers that are taking care about recursive references: **AdaptedObjectSerializer** and **UserSerializerSerializer**. SerializerBuilder#build() is the class that will give the required serializer.

**AdaptedObjectSerializer** is activated when you specify a custom adapter. For example:
```
Jsonb adapterSerializerJsonb = JsonbBuilder.create(new JsonbConfig()
            .withAdapters(new ChainAdapter(), new FooAdapter()));
```
**UserSerializerSerializer** is activated when you specify a custom serializer, for example:
```
Jsonb userSerializerJsonb = JsonbBuilder.create(new JsonbConfig()
            .withSerializers(new ChainSerializer(), new FooSerializer()));
```
But note that when any of previous serializers are activated, the **ObjectSerializer** will not be activated. This is because your custom serializer/adapter is going to replace ObjectSerializer.

In conclusion, AdaptedObjectSerializer and UserSerializerSerializer could interfere with ObjectSerializer because they share the same context (same HashSet). But in practice it will not happen.

I wanted to test this also with the UserSerializerSerializer but it is not possible (or I don't know how to do it). It complains because generator.writeStartObject() is invoked 2 times before it is closed. I'm getting: 

> Caused by: javax.json.stream.JsonGenerationException: Illegal method during JSON generation, not valid in current context IN_OBJECT
> 	at org.glassfish.json.JsonGeneratorImpl.writeStartObject(JsonGeneratorImpl.java:121)
> 	at org.eclipse.yasson.adapters.model.FooSerializer.serialize(FooSerializer.java:11)
> 	at org.eclipse.yasson.adapters.model.FooSerializer.serialize(FooSerializer.java:1)
> 	at org.eclipse.yasson.internal.serializer.UserSerializerSerializer.serialize(UserSerializerSerializer.java:54)
> 	at org.eclipse.yasson.internal.Marshaller.serializeRoot(Marshaller.java:147)
> 	at org.eclipse.yasson.internal.Marshaller.serialize(Marshaller.java:126)
> 	at org.eclipse.yasson.adapters.model.ChainSerializer.serialize(ChainSerializer.java:18)
> 	at org.eclipse.yasson.adapters.model.ChainSerializer.serialize(ChainSerializer.java:1)
> 	at org.eclipse.yasson.internal.serializer.UserSerializerSerializer.serialize(UserSerializerSerializer.java:54)
> 	at org.eclipse.yasson.internal.Marshaller.serializeRoot(Marshaller.java:147)
> 	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:75)

With the next code:
```
private static final Jsonb userSerializerJsonb = JsonbBuilder.create(new JsonbConfig()
            .withSerializers(new ChainSerializer(), new FooSerializer()));
```

```
public class ChainSerializer implements JsonbSerializer<Chain>{
    
    @Override
    public void serialize(Chain obj, JsonGenerator generator, SerializationContext ctx) {
        
        Marshaller context = (Marshaller)ctx;
        generator.writeStartObject();
        generator.write("name", obj.getName());
        if(obj.getHas() != null) {
            context.serialize(obj.getHas(), generator);
        }
        if(obj.getLinksTo() != null) {
            context.serialize(obj.getLinksTo(), generator);
        }
        generator.writeEnd();
        
    }
    
}
```

```
public class FooSerializer implements JsonbSerializer<Foo>{
    
    @Override
    public void serialize(Foo obj, JsonGenerator generator, SerializationContext ctx) {
        generator.writeStartObject();
        generator.write("bar", obj.getBar());
        generator.writeEnd();
    }
    
}
```

Reply:
Finally I wrote a recursive Serializer that works. I don't get the error:
javax.json.stream.JsonGenerationException: Illegal method during JSON generation, not valid in current context IN_OBJECT

It seems you cannot create your own serializer that delegates in other in an easy way. I will probably create a different issue to take a look at this.

Reply:
@aguibert after the tests I have added do you think it is safe to be merged?.